### PR TITLE
[Popover] Set right position when preferred alignment is right.

### PR DESCRIPTION
**Pull request by dleroux**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/polaris-react/issues/1673

### WHAT is this pull request doing?

From what I can tell the issue is that the Scrollable scrollbars appear inside the Popover after the PositionOverlay has done its calculations and since the content is responsive, once the scrollbars are added the content wraps. 

For example. If the window is 1000px wide, the overlay 100px wide, the left position will be calculated at 900px. Once the scrollbars appear, the left position does not get recalculated because the content is responsive. The left position pushed the overlay against the right side of the browser and when too close to the chrome causes the content to wrap. This occurs only when the popover is too close to the right of the window. 

This is one of a few attempts at fixing this issue. When a popover has a preferred right, we go from the right instead of the left. That way, even if scrollbars appear, the right position stays the same, and the left does not counter it. 

It also contains a change that when content can be scrolled in a Scrollable container, it sets `overflow-y: scroll` then reverts to `auto` when the content cannot be scrolled. This forces the layout engines to set the element width properly.

## 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React, {useState} from 'react';
import {ArrowLeftMinor} from '@shopify/polaris-icons';
import {
  AppProvider,
  Frame,
  Page,
  Card,
  ActionList,
  TopBar,
  Button,
  Popover,
} from '../src';

export function Playground() {
  const [userMenuOpen, setUserMenuOpen] = useState(false);
  const [searchActive, setSearchActive] = useState(false);
  const [searchText, setSearchText] = useState('');
  const [popover1Active, setPopover1Active] = useState(false);
  const [popover2Active, setPopover2Active] = useState(false);
  const [popover3Active, setPopover3Active] = useState(false);
  const [popover1PositionBelow, setPopover1PositionBelow] = useState(true);
  const [popover2PositionBelow, setPopover2PositionBelow] = useState(true);
  const [popover3PositionBelow, setPopover3PositionBelow] = useState(true);
  const [showScrollingUserMenu, setShowScrollingUserMenu] = useState(true);

  const toggleUserMenu = () => {
    setUserMenuOpen(!userMenuOpen);
  };

  const toggleScrollingUserMenu = () => {
    setShowScrollingUserMenu(!showScrollingUserMenu);
  };

  const togglePopover = (popoverID: string) => () => {
    setPopover1Active(popoverID === 'popover1' ? !popover1Active : false);
    setPopover2Active(popoverID === 'popover2' ? !popover2Active : false);
    setPopover3Active(popoverID === 'popover3' ? !popover3Active : false);
  };

  const handleSearchResultsDismiss = () => {
    setSearchActive(false);
    setSearchText('');
  };

  const handleSearchChange = (value) => {
    setSearchText(value);
    if (value.length > 0) {
      setSearchActive(true);
    } else {
      setSearchActive(false);
    }
  };

  const theme = {
    colors: {
      topBar: {
        background: '#357997',
        backgroundLighter: '#6192a9',
        color: '#FFFFFF',
      },
    },
    logo: {
      width: 124,
      topBarSource:
        'https://cdn.shopify.com/s/files/1/0446/6937/files/jaded-pixel-logo-color.svg?6215648040070010999',
      url: 'http://jadedpixel.com',
      accessibilityLabel: 'Jaded Pixel',
    },
  };

  const actionsWithScroll = [
    {
      items: [{content: 'Back to Shopify', icon: ArrowLeftMinor}],
    },
    {
      items: [{content: 'Community forums'}],
    },
    {
      items: [{content: 'Community forums'}],
    },
    {
      items: [{content: 'Community forums'}],
    },
    {
      items: [{content: 'Community forums'}],
    },
    {
      items: [{content: 'Community forums'}],
    },
    {
      items: [{content: 'Community forums'}],
    },
  ];

  const actions = [
    {
      items: [{content: 'Back to Shopify', icon: ArrowLeftMinor}],
    },
    {
      items: [{content: 'Community forums'}],
    },
    {
      items: [{content: 'Community forums'}],
    },
    {
      items: [{content: 'Community forums'}],
    },
  ];

  const userMenuMarkup = (
    <TopBar.UserMenu
      actions={showScrollingUserMenu ? actionsWithScroll : actions}
      name="Dharma"
      detail="Jaded Pixel"
      initials="D"
      open={userMenuOpen}
      onToggle={toggleUserMenu}
    />
  );

  const searchResultsMarkup = (
    <Card>
      <ActionList
        items={[
          {content: 'Shopify help center'},
          {content: 'Community forums'},
        ]}
      />
    </Card>
  );

  const searchFieldMarkup = (
    <TopBar.SearchField
      onChange={handleSearchChange}
      value={searchText}
      placeholder="Search"
    />
  );

  const topBarMarkup = (
    <TopBar
      showNavigationToggle
      userMenu={userMenuMarkup}
      searchResultsVisible={searchActive}
      searchField={searchFieldMarkup}
      searchResults={searchResultsMarkup}
      onSearchResultsDismiss={handleSearchResultsDismiss}
    />
  );

  const activator1 = (
    <Button onClick={togglePopover('popover1')}>Popover Left</Button>
  );

  const activator2 = (
    <Button onClick={togglePopover('popover2')}>Popover Center</Button>
  );

  const activator3 = (
    <Button onClick={togglePopover('popover3')}>Popover Right</Button>
  );

  return (
    <AppProvider i18n={{}} theme={theme}>
      <Frame topBar={topBarMarkup}>
        <Page
          title="Popover alignment/width bug"
          primaryAction={{
            content: `Render ${
              showScrollingUserMenu ? 'non-scrolling' : 'scrolling'
            } user menu`,
            onAction: toggleScrollingUserMenu,
          }}
        >
          <Card>
            <Card.Section
              actions={[
                {
                  content: 'Toggle position',
                  onAction: () =>
                    setPopover1PositionBelow(!popover1PositionBelow),
                },
              ]}
            >
              <Popover
                active={popover1Active}
                activator={activator1}
                onClose={togglePopover('popover1')}
                preferredAlignment="left"
                preferredPosition={popover1PositionBelow ? 'below' : 'above'}
              >
                <ActionList
                  items={[{content: 'Import'}, {content: 'Export'}]}
                />
              </Popover>
            </Card.Section>
            <Card.Section
              actions={[
                {
                  content: 'Toggle position',
                  onAction: () =>
                    setPopover2PositionBelow(!popover2PositionBelow),
                },
              ]}
            >
              <Popover
                active={popover2Active}
                activator={activator2}
                onClose={togglePopover('popover2')}
                preferredAlignment="center"
                preferredPosition={popover2PositionBelow ? 'below' : 'above'}
              >
                <ActionList
                  items={[{content: 'Import'}, {content: 'Export'}]}
                />
              </Popover>
            </Card.Section>
            <Card.Section
              actions={[
                {
                  content: 'Toggle position',
                  onAction: () =>
                    setPopover3PositionBelow(!popover3PositionBelow),
                },
              ]}
            >
              <Popover
                active={popover3Active}
                activator={activator3}
                onClose={togglePopover('popover3')}
                preferredAlignment="right"
                preferredPosition={popover3PositionBelow ? 'below' : 'above'}
              >
                <ActionList
                  items={[{content: 'Import'}, {content: 'Export'}]}
                />
              </Popover>
            </Card.Section>
          </Card>
        </Page>
      </Frame>
    </AppProvider>
  );
}

```

</details>

### 🎩 checklist

* [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [x] Updated the component's `README.md` with documentation changes
* [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [ ] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the Polaris UI kit

### CODE REVIEW

File: src/components/Popover/Popover.scss [hunk 1]
```diff
@@ -52,7 +52,7 @@ $content-max-width: rem(400px);
 }
 
 .positionedAbove {
-  margin: spacing() 0 $visible-portion-of-arrow spacing(tight);
+  margin: spacing() spacing(tight) $visible-portion-of-arrow;
```
Comment (dleroux) [inline src/components/Popover/Popover.scss]: Not sure why we needed to remove the right margin when positioned above.

File: src/components/PositionedOverlay/PositionedOverlay.tsx [hunk 2]
```diff
@@ -234,7 +250,10 @@ export class PositionedOverlay extends React.PureComponent<
           {
             measuring: false,
             activatorRect: getRectForNode(activator),
-            left: horizontalPosition,
+            left:
+              preferredAlignment !== 'right' ? horizontalPosition : undefined,
+            right:
```
Comment (dleroux) [inline src/components/PositionedOverlay/PositionedOverlay.tsx]: We're only going to go from the right if preferred alignment is right.

File: src/components/PositionedOverlay/PositionedOverlay.tsx [hunk 3]
```diff
@@ -298,7 +317,7 @@ function windowRect() {
     top: window.scrollY,
     left: window.scrollX,
     height: window.innerHeight,
-    width: window.innerWidth,
+    width: document.body.clientWidth,
```
Comment (dleroux) [inline src/components/PositionedOverlay/PositionedOverlay.tsx]: window.innerWidth includes the scrollbars, We don't want this when aligning from the right and it's shouldn't impact the left.

File: src/components/PositionedOverlay/utilities/math.ts [hunk 4]
```diff
@@ -98,13 +98,11 @@ export function calculateHorizontalPosition(
       Math.max(0, activatorRect.left - overlayMargins.horizontal),
     );
   } else if (preferredAlignment === 'right') {
-    const activatorRight = activatorRect.left + activatorRect.width;
+    const activatorRight =
+      containerRect.width - (activatorRect.left + activatorRect.width);
```
Comment (dleroux) [inline src/components/PositionedOverlay/utilities/math.ts]: the right position calculation.

File: src/components/TopBar/components/Menu/Menu.tsx [hunk 5]
```diff
@@ -57,6 +57,7 @@ export function Menu(props: MenuProps) {
       onClose={onClose}
       fixed
       fullHeight={isFullHeight}
+      preferredAlignment="right"
```
Comment (dleroux) [inline src/components/TopBar/components/Menu/Menu.tsx]: The user menu was causing the issue. This fixed it.

File: src/components/PositionedOverlay/PositionedOverlay.tsx (hunk 3 above)
Comment (kyledurand) [reply src/components/PositionedOverlay/PositionedOverlay.tsx]: TIL! Will this change alone fix the measuring then and the other stuff in this PR is just a nice to have?

Comment (dleroux) [reply src/components/PositionedOverlay/PositionedOverlay.tsx]: No, the right positioning is what is needed. I'll write it in the description.

File: src/components/PositionedOverlay/PositionedOverlay.tsx [hunk 6]
```diff
@@ -278,9 +298,9 @@ export function intersectionWithViewport(
 function getMarginsForNode(node: HTMLElement) {
   const nodeStyles = window.getComputedStyle(node);
   return {
-    activator: parseFloat(nodeStyles.marginTop || ''),
-    container: parseFloat(nodeStyles.marginBottom || ''),
-    horizontal: parseFloat(nodeStyles.marginLeft || ''),
+    activator: parseFloat(nodeStyles.marginTop || '0'),
+    container: parseFloat(nodeStyles.marginBottom || '0'),
+    horizontal: parseFloat(nodeStyles.marginLeft || '0'),
```
Comment (dleroux) [inline src/components/PositionedOverlay/PositionedOverlay.tsx]: defaulting to 0. an empty string didn't make sense for the tests.

File: src/components/PositionedOverlay/tests/PositionedOverlay.test.tsx [hunk 7]
```diff
@@ -70,6 +70,20 @@ describe('<PositionedOverlay />', () => {
...
+    it('aligns right if preferredAlignment is given', () => {
+      const positionedOverlay = mountWithAppProvider(
+        <PositionedOverlay {...mockProps} preferredAlignment="right" />,
+      );
+
+      expect((positionedOverlay.find('div').prop('style') as any).right).toBe(
+        0,
+      );
       expect(
         (positionedOverlay.find('div').prop('style') as any).left,
```
Comment (dleroux) [inline src/components/PositionedOverlay/tests/PositionedOverlay.test.tsx]: This test didn't make sense. No matter whar left would be undefined, which it shouldn't be.

File: UNRELEASED.md [hunk 8]
```diff
@@ -7,6 +7,7 @@
 ### Bug fixes
 
 - Fixed issue with `Filters` component displaying an undesired margin top and bottom on the button element on Safari ([#2292](https://github.com/Shopify/polaris-react/pull/2292))
+- Fixed `PositionedOverlay` not correctly getting its position when aligned to the right of activator ([#2587](https://github.com/Shopify/polaris-react/pull/2587))
```
Comment (dleroux) [inline UNRELEASED.md]:
```suggestion
- Fixed `PositionedOverlay` not correctly getting its position when aligned to the right of the activator ([#2587](https://github.com/Shopify/polaris-react/pull/2587))
```

**Comment by dleroux** [comment]
:( Just saw that safari still wraps for no good reason.

**Comment by dleroux** [comment]
Looks like this approach works perfectly on Chrome, but Firefox and Safari do not resize the parent `PositionOverlay` when the scrollbars appear. I was hoping a simple CSS change would fix it but I was unsuccessful.

**Comment by chloerice** [comment]
> Looks like this approach works perfectly on Chrome, but Firefox and Safari do not resize the parent `PositionOverlay` when the scrollbars appear. I was hoping a simple CSS change would fix it but I was unsuccessful.

Yeah, been there with these changes 😞 Every approach we tried was inconsistent across browsers. Only idea I have left is to start from scratch with this component, I feel like a lot of the logic is unnecessary, but because it's unclear what some of it is meant to do, refactoring it while trying to fix this just kept breaking it in other ways 😣.

**Comment by dleroux** [comment]
@chloerice I thought of redoing it but after wrapping my head around it it's actually pretty good and I fear we'll end up in the same spot. The only thing I can think of is to add complexity and measure whether or not the content will cause scrollbars, then measure the scrollbar, then add the scrollbar width to the container. The math is all correct.

I asked @tmlayton to take a look and see if he can see something we're missing.

It works perfectly in Chrome right now. It's almost tempting to ship this.

**Comment by AndrewMusgrave** [comment]
> It's almost tempting to ship this.

We could always revert this when we have a better fix. Getting something out that'll work for the majority of consumers seems like a win to me.

**Comment by alex-page** [comment]
I made this PR https://github.com/Shopify/polaris-react/pull/2362 that starts the logic again from scratch, removing the confusing `Math.max`. 

As Tim mentioned it needs tests to make sure it doesn't break anything.

**Comment by tmlayton** [comment]
Alright, this should be good now. When content can be scrolled in a `Scrollable` container, it sets `overflow-y: scroll` and reverts to `auto` when the content cannot be scrolled. This forces the layout engines to set the element width properly. Here are screenshots from Chrome, FF, Safari, and Edge with and without scrollbars:

### Always on scrollbars
Screen Shot 2020-01-10 at 4 31 05 PM

### Always on scrollbars with non-scrolling content
Screen Shot 2020-01-10 at 4 31 25 PM

### Floating scrollbars (does not apply to Edge)
Screen Shot 2020-01-10 at 4 32 37 PM

**Comment by dleroux** [comment]
@tmlayton you are a 🌟 🙇

**Pull request merged**